vms/proposervm/proposer/windower.go

```diff
@@ -70,32 +75,71 @@ func New(state validators.State, subnetID, chainID ids.ID) Windower {
 	}
 }
 
-func (w *windower) Proposers(ctx context.Context, chainHeight, pChainHeight uint64, maxWindows int) ([]ids.NodeID, error) {
-	// get the validator set by the p-chain height
-	validatorsMap, err := w.state.GetValidatorSet(ctx, pChainHeight, w.subnetID)
+func (w *windower) Delay(ctx context.Context, chainHeight, pChainHeight uint64, validatorID ids.NodeID, maxWindows int) (time.Duration, error) {
+	if validatorID == ids.EmptyNodeID {
+		return time.Duration(maxWindows) * WindowDuration, nil
+	}
+
+	proposers, err := w.proposers(ctx, chainHeight, pChainHeight, maxWindows)
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 
-	// convert the map of validators to a slice
-	validators := make([]validatorData, 0, len(validatorsMap))
-	weight := uint64(0)
-	for k, v := range validatorsMap {
-		validators = append(validators, validatorData{
-			id:     k,
-			weight: v.Weight,
-		})
-		newWeight, err := math.Add64(weight, v.Weight)
-		if err != nil {
-			return nil, err
+	delay := time.Duration(0)
+	for _, nodeID := range proposers {
+		if nodeID == validatorID {
+			return delay, nil
 		}
-		weight = newWeight
+		delay += WindowDuration
 	}
+	return delay, nil
+}
 
-	// canonically sort validators
-	// Note: validators are sorted by ID, sorting by weight would not create a
-	// canonically sorted list
-	utils.Sort(validators)
+func (w *windower) ExpectedProposer(
+	ctx context.Context,
+	chainHeight,
+	pChainHeight uint64,
+	blockTime,
+	parentBlockTime time.Time,
+) (ids.NodeID, error) {
+	validators, _, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	// convert the slice of validators to a slice of weights
+	validatorWeights := make([]uint64, len(validators))
+	for i, v := range validators {
+		validatorWeights[i] = v.weight
+	}
+
+	if err := w.sampler.Initialize(validatorWeights); err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	var (
+		numToSample = 1
+		slot        = uint32(blockTime.Sub(parentBlockTime) / WindowDuration)
```

Comment:
Is there any particular reason why we use `uint32` for this (other than that we never expect to have that many slots)?

Reply:
Apparently only the lower 32bits are used to seed. See [here](https://github.com/gonum/gonum/blob/82e11dfec82f254ff49e0aaafd9ba59736542a35/mathext/prng/mt19937.go#L39).

There are UTs failing  if we use the full uint64.
@StephenButtolph could that be the reason why we didn't use uint64 before?

Reply:
(as Stephen discovered, we can use the same library: https://github.com/gonum/gonum/blob/82e11dfec82f254ff49e0aaafd9ba59736542a35/mathext/prng/mt19937_64.go#L35)